### ISSUE TITLE
New mutable hash maps no longer expose elemEquals/elemHashCode extension point

### ISSUE THREAD

**Issue by retronym** [scala/bug#11370]
This is a consequence of the change to compose `HashTable` rather than extend it.

If this is an intentional breaking change, we should go a bit further and inline calls to `elemEquals` and `elemHashCode` with calls to `==` and `##`.

**Comment by szeiger** [comment]
We could offer the same methods with the same names for users to override in order to make it source compatible. In the end it comes down to performance. How expensive are calls to non-final methods in the most performance-critical sections? I'll run some benchmarks.

**Comment by szeiger** [comment]
No difference in peformance. Apparently the small hot methods always get inlined by HotSpot.

But the class is already marked `@deprecatedInheritance` because we want to disallow extending concrete collection implementations. Should we revert this decision or recommend wrapping the keys?

**Comment by shanielh** [comment]
I always liked .net approach to this: passing a trait that has elemHashCode and elemEquals (They called it IEqualityComparer) instead of inherit the collection type. 

I guess it’s too late now :-)

**Comment by retronym** [comment]
I think we should leave then final for now.

I'm surprised that the benchmark couldn't show a difference. The benchmark would need to take some care to make things realistically hard for HotSpot.

**Comment by szeiger** [comment]
I agree. Configurable equality / hashing with a typeclass would be a nice feature. I couldn't find an existing ticket for it so let's use this one.
